### PR TITLE
feat: Introducing GenericKubernetesResource

### PR DESCRIPTION
**Pull request by manusa**
### Introduction of the GenericKubernetesResource

This class holds the minimum structure for any valid resource served by the Kubernetes API. i.e. A HasMetadata implementation with an `ObjectMeta metadata` field and a Map for any additional property/field.

The main purpose of this class is to be able to Serialize/Deserialize any resource served by Kubernetes regardless of a specific type previously registered for that resource.

The Serializer now looks for any registered type for a given resource and in case none is found, the resource is deserialized as a GenericKubernetesResource. The immediate benefit is that deserialization will never fail.

This resource can be later on used in a specific [`OperationSupport`](https://github.com/fabric8io/kubernetes-client/blob/71c1a8c09035f09d65c16d7fcface9e509518a52/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/OperationSupport.java) implementation that will provide **consistent** operation support for any resource.

### RawCustomResourceOperationsImpl uses GenericKubernetesResourceOperationsImpl

This brings consistency to the client, but no specific behavioral changes should be noticed. Some changes might be noticed in case of the MockServer and white box testing, since some of the requests have changed, but the result should be no different.

The only major difference is that a `.get()` for a non-existent resource previously threw an Exception and now returns null (just like the rest of OperationSupport implementations for any other resource).

Redundant methods have been marked as deprecated. We should consider removing them definitely after the next few releases.

### Next steps

Improve `KubernetesClient#load` to support operations for GenericKubernetesResource. This means requesting the applicable CustomResourceDefinition from the cluster in order to infer the plural and scope of the managed resource.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by shawkins** [comment]
This change will be needed to rework/remove the raw websocket watch logic - it seems broken in the current state because it doesn't fully understand the events (http gone, updating the resource version, lists, etc.).

**Comment by shawkins** [comment]
Should we hold off on #3156 until after this work?  If so, I will split off the additional logic that addresses #3194 as a separate pr.

**Comment by manusa** [comment]
> Should we hold off on #3156 until after this work? If so, I will split off the additional logic that addresses #3194 as a separate pr.

My intention is to at least finish the initial work on this by today (I already spent 2 days working on it and can't spend more time).

This PR introduces the Generic Resource that can be further used in many areas of the client but those should be tackled separately.

**Comment by shawkins** [comment]
> This PR introduces the Generic Resource that can be further used in many areas of the client but those should be tackled separately.

Okay, I'll hold off on adopting this for now in the mock.  In the meantime I've rebased #3156 to master, when you get a chance please review.

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java [hunk 1]
```diff
@@ -82,7 +82,7 @@ private static KubernetesResource fromObjectNode(JsonParser jp, JsonNode node) t
         if (key != null) {
```
Comment (shawkins) [inline kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java:82]: I tried these changes with removing FallbackHasMetadata.  Some of the raw logic was written with tests that do not specify the kind in a create.  I'm not sure if that's allowed, but we could easily account for that here by using the generic logic when either the key or the resourceType is null.

**Review by shawkins — CHANGES_REQUESTED** [review]

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java (hunk 1 above)
Comment (manusa) [reply kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java:82]: I cannot think of a case (besides JKube's Fragments) where you can define a resource and not specify the kind and the apiVersion.

We could modify to always try to deserialize to GenericKubernetesResource and let Jackson fail. However I'm not sure of the implications or backwards compatibility side-effects of this change:

```java
    private static KubernetesResource fromObjectNode(JsonParser jp, JsonNode node) throws IOException {
        String key = getKey(node);
        if (key != null) {
            Class<? extends KubernetesResource> resourceType = mapping.getForKey(key);
            if (resourceType != null && KubernetesResource.class.isAssignableFrom(resourceType)){
                return jp.getCodec().treeToValue(node, resourceType);
            }
        }
        return jp.getCodec().treeToValue(node, GenericKubernetesResource.class);
    }
```

Comment (shawkins) [reply kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/internal/KubernetesDeserializer.java:82]: After trying to modify test-rawcustomresource.yml, I see kind is definitely required.  So CustomResourceCrudTest and a unit test that was added in the mock logic need to be updated to include kind to incorporate your changes.  With that understanding what you had to start with is good.

File: kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/GenericKubernetesResource.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class GenericKubernetesResource implements HasMetadata {
+
+  @JsonProperty("apiVersion")
+  private String apiVersion;
+  @JsonProperty("kind")
+  private String kind;
+  @JsonProperty("metadata")
+  private ObjectMeta metadata;
+  @JsonIgnore
+  private transient Map<String, Object> additionalProperties = new HashMap<>();
```
Comment (shawkins) [inline kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/GenericKubernetesResource.java:54]: Can we use a LinkedHashMap here instead?  For the mock server logic I'll be able to manipulate GenericKubernetesResource objects directly and that will keep the output json looking very similar to the input.

Comment (manusa) [reply kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/GenericKubernetesResource.java:54]: I think we can set the default value to LinkedHashMap.
However, I'm not sure if we can guarantee the same for deserialization purposes.

Comment (shawkins) [reply kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/GenericKubernetesResource.java]: I tested locally using 
private transient LinkedHashMap<String, Object> additionalProperties = new LinkedHashMap<>(); 
and everything worked as expected.

**Review by shawkins — APPROVED** [review]
LGTM

**Comment by manusa** [comment]
Hi @lburgazzoli, since (I think) you are an active user of the RawCustomResources, it would be very interesting to count with your review and check in case something is broken for you with the changes suggested in this PR. :pray:

**Comment by lburgazzoli** [comment]
@manusa I'll try to build the PR and use it in my project beginning of next week, my knowledge of the k8s client internals is too limited to give any valuable review. If there is an example that show how to use the new resource, I may take a look at it today to see if I spot something missing.

**Comment by manusa** [comment]
> If there is an example that show how to use the new resource, I may take a look at it today to see if I spot something missing.

Basically there is no way to explicitly use the new Resource (as of now). The partial purpose of this PR RawCustomResourceOperationsImpl instead of re-implementing everything from scratch (and leading to issues like some of the ones you reported (#3023, and indirectly #2929)) it will use the regular implementations by delegating the logic to the new `GenericKubernetesResourceOperationsImpl`.

So if you're still using the "unstructured" RawCustomResourceOperationsImpl ([`kubernetesClient.customResource(CustomResourceDefinitionContext)`](https://github.com/fabric8io/kubernetes-client/blob/c0a0107e2420fad3d2ad9c76028066401364536e/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java#L206)), the new functionality will kick in. Any troubles you may have had previously with this DSL should remain fixed, and functionality should be complete.

This is also the base to any enhancement or feature we want to perform regarding #2929.

**Comment by lburgazzoli** [comment]
ok, so I'll try to compile it and see if it breaks something

**Review by rohanKanojia — APPROVED** [review]

**Comment by manusa** [comment]
> @manusa I've updated my operator to use the code from this PR and nothing seems to be broken, so LGTM

Great, thx for trying it out. Merging :rocket:

**Pull request merged**